src/wp-includes/functions.php

```diff
@@ -7474,17 +7474,34 @@ function get_tag_regex( $tag ) {
  *
  * @see https://core.trac.wordpress.org/ticket/23688
  *
- * @param string $charset A charset name.
+ * @param string $charset A charset name, e.g. "UTF-8", "Windows-1252", "SJIS".
  * @return string The canonical form of the charset.
  */
 function _canonical_charset( $charset ) {
-	if ( 'utf-8' === strtolower( $charset ) || 'utf8' === strtolower( $charset ) ) {
-
+	if ( is_utf8_charset( $charset ) ) {
 		return 'UTF-8';
 	}
 
-	if ( 'iso-8859-1' === strtolower( $charset ) || 'iso8859-1' === strtolower( $charset ) ) {
-
+	/*
+	 * Normalize the ISO-8859-1 family of languages.
+	 *
+	 * This is not required for htmlspecialchars(), as it properly recognizes all of
+	 * the input character sets that here are transformed into "ISO-8859-1".
+	 *
+	 * @todo Should this entire check be removed since it's not required for the stated purpose?
+	 * @todo Should WordPress transform other potential charset equivalents, such as "latin1"?
+	 */
+	$charset_length = strlen( $charset );
+	if (
+		(
+			10 === $charset_length &&
+			substr_compare( $charset, 'iso-8859-1', 0, 10, true )
```

Comment:
This function has an [interesting signature](https://www.php.net/manual/en/function.substr-compare.php) and it appears to be misused here:

```
substr_compare(
    string $haystack,
    string $needle,
    int $offset,
    ?int $length = null,
    bool $case_insensitive = false
): int
```

Return Values
Returns -1 if haystack from position offset is less than needle, 1 if it is greater than needle, and 0 if they are equal. If offset is equal to (prior to PHP 7.2.18, 7.3.5) or greater than the length of haystack, or the length is set and is less than 0, substr_compare() prints a warning and returns false.

The needle and haystack appear to be swapped, although that's probably irrelevant when they're the same length. However, this should not be a truthy check but a check that `0 === substr_compare(…)`, we expect the needle to be found at offset 0.

Reply:
thanks @sirreal - absolutely correct on the return value.

also the needle and haystack aren't swapped here - that's just how I wrote this. as you noted, given that they are the same length it doesn't matter.

I'm preferring this function with the goal of avoiding needless string allocation and processing with `strtolower()`